File: gtk-synth/reference/modules/interface/strategies/VaporStrategy.js
// VaporStrategy.js
// Vapor interface: Chaotic aeroacoustic turbulence

import { InterfaceStrategy } from '../InterfaceStrategy.js';
import { ChaoticOscillator } from '../utils/ExcitationGen.js';
import { softClip } from '../utils/NonlinearityLib.js';

export class VaporStrategy extends InterfaceStrategy {
    constructor(sampleRate) {
        super(sampleRate);

        // Multiple chaotic oscillators with different seeds
        this.chaos1 = new ChaoticOscillator(3.7);
        this.chaos2 = new ChaoticOscillator(3.8);
        this.chaos3 = new ChaoticOscillator(3.9);

        // Previous outputs for feedback
        this.prev1 = 0;
        this.prev2 = 0;
    }

    /**
     * Process one sample through vapor interface
     * Chaotic turbulence with feedback coupling
     * @param {number} input - Input sample
     * @returns {number} Processed output
     */
    process(input) {
        // Map intensity to chaos parameter (3.57+ = chaotic)
        const r = 2.5 + this.intensity * 1.5;

        this.chaos1.setR(r);
        this.chaos2.setR(r + 0.1);
        this.chaos3.setR(r + 0.2);

        // Generate chaotic signals
        const c1 = this.chaos1.process(0.3);
        const c2 = this.chaos2.process(0.3);
        const c3 = this.chaos3.process(0.3);

        // Mix input with chaotic forcing
        const chaosAmount = this.intensity * 0.6;
        const inputAmount = 1 - chaosAmount * 0.5;

        const mixed = input * inputAmount + (c1 + c2 + c3) * chaosAmount;

        // Couple with feedback from previous samples
        const feedback = (this.prev1 * 0.3 + this.prev2 * 0.2) * chaosAmount;

        const turbulent = mixed + feedback;

        // Soft clip to prevent runaway
        const output = softClip(turbulent, 1.2);

        // Store for feedback
        this.prev2 = this.prev1;
        this.prev1 = output;

        return Math.max(-1, Math.min(1, output));
    }

    /**
     * Reset state
     */
    reset() {
        this.chaos1.reset();
        this.chaos2.reset();
        this.chaos3.reset();
        this.prev1 = 0;
        this.prev2 = 0;
    }

    /**
     * Called on note-on trigger
     */
    onNoteOn() {
        this.reset();
    }
}

File: flues-synth/src/audio/modules/strategies/vapor_strategy.c
// vapor_strategy.c
// Vapor interface: Chaotic aeroacoustic turbulence
// Translated from reference/modules/interface/strategies/VaporStrategy.js

#include "interface_strategy.h"
#include "dsp_utils.h"
#include <stdlib.h>
#include <math.h>

typedef struct {
    ChaoticOscillator chaos1;
    ChaoticOscillator chaos2;
    ChaoticOscillator chaos3;
    float prev1;
    float prev2;
} VaporImpl;

static float vapor_process(InterfaceStrategy* self, float input) {
    VaporImpl* impl = (VaporImpl*)self->impl_data;

    // Map intensity to chaos parameter (3.57+ = chaotic)
    const float r = 2.5f + self->intensity * 1.5f;

    chaotic_oscillator_set_r(&impl->chaos1, r);
    chaotic_oscillator_set_r(&impl->chaos2, r + 0.1f);
    chaotic_oscillator_set_r(&impl->chaos3, r + 0.2f);

    // Generate chaotic signals
    const float c1 = chaotic_oscillator_process(&impl->chaos1, 0.3f);
    const float c2 = chaotic_oscillator_process(&impl->chaos2, 0.3f);
    const float c3 = chaotic_oscillator_process(&impl->chaos3, 0.3f);

    // Mix input with chaotic forcing
    const float chaos_amount = self->intensity * 0.6f;
    const float input_amount = 1.0f - chaos_amount * 0.5f;

    const float mixed = input * input_amount + (c1 + c2 + c3) * chaos_amount;

    // Couple with feedback from previous samples
    const float feedback = (impl->prev1 * 0.3f + impl->prev2 * 0.2f) * chaos_amount;

    const float turbulent = mixed + feedback;

    // Soft clip to prevent runaway
    const float output = soft_clip_drive(turbulent, 1.2f);

    // Store for feedback
    impl->prev2 = impl->prev1;
    impl->prev1 = output;

    return fmaxf(-1.0f, fminf(1.0f, output));
}

static void vapor_reset(InterfaceStrategy* self) {
    VaporImpl* impl = (VaporImpl*)self->impl_data;
    chaotic_oscillator_reset(&impl->chaos1);
    chaotic_oscillator_reset(&impl->chaos2);
    chaotic_oscillator_reset(&impl->chaos3);
    impl->prev1 = 0.0f;
    impl->prev2 = 0.0f;
}

static void vapor_set_intensity(InterfaceStrategy* self, float intensity) {
    self->intensity = intensity;
}

static void vapor_set_gate(InterfaceStrategy* self, bool gate) {
    if (gate) {
        vapor_reset(self);
    }
}

static void vapor_destroy(InterfaceStrategy* self) {
    if (self) {
        free(self->impl_data);
        free(self);
    }
}

static InterfaceStrategyVTable vapor_vtable = {
    .process = vapor_process,
    .reset = vapor_reset,
    .set_intensity = vapor_set_intensity,
    .set_gate = vapor_set_gate,
    .destroy = vapor_destroy
};

InterfaceStrategy* vapor_strategy_create(float sample_rate) {
    (void)sample_rate;  // Unused
    InterfaceStrategy* strategy = (InterfaceStrategy*)calloc(1, sizeof(InterfaceStrategy));
    if (!strategy) return NULL;

    VaporImpl* impl = (VaporImpl*)calloc(1, sizeof(VaporImpl));
    if (!impl) {
        free(strategy);
        return NULL;
    }

    chaotic_oscillator_init(&impl->chaos1, 3.7f);
    chaotic_oscillator_init(&impl->chaos2, 3.8f);
    chaotic_oscillator_init(&impl->chaos3, 3.9f);
    impl->prev1 = 0.0f;
    impl->prev2 = 0.0f;

    strategy->vtable = &vapor_vtable;
    strategy->impl_data = impl;
    strategy->intensity = 0.5f;

    return strategy;
}
